Silent token renewal

ciaranj opened the pull request:
HI there,

I've done some work to extend your directive to support silent token renewal. The approach leans a little (the renewTokenSilently method) on the approach that ThinkTecture provide in their OIDC sample libraries (https://github.com/IdentityServer/IdentityServer3.Samples/blob/master/source/Clients/JavaScriptImplicitClient/oidc.js) which is Apache licensed, since you framed this library as being a client for IdentityServerv3 originally I figured that would be ok?

ciaranj committed:
Check for expired tokens on route change.

* The absence of a token should be considered the same as having a token
  but it being expired.  In both scenarios now on a route change the user
  will be redirected to the IDP

JamesRandall commented:
Thanks for this - and apologies it's taken me a while to pick up on it. Been on a bit of a break. I'm just making a couple of simple tweaks at the moment and when that's done I'll look to merge this in.

ciaranj commented:
No worries ;)

ciaranj commented:
Any joy with this :)

JamesRandall commented:
Hi - just not had a chance yet as pretty busy with the day job. I've merged a couple of minor fixes but this is a bit more substantial!

ciaranj commented:
;) I understand, the PRs just keep backing up on my own projects too!

Commit:
Provide 'silent' token-renewal support

Silent renewal of tokens is achieved by the dynamic construction of a
child iframe that attempts to re-authenticate the page  with the same
original arguments as the 'parent' page, but with an additional prompt=none
argument.  If this successfully responds with a token we update the token
we're sending for API calls to this renewed value, otherwise we sign-out.

The specific details of how this is achieved within this plugin are mildly
complex so i'll describe them here.

Parent lifetime:
  If the parent receives an 'oauth2:authSuccess' event and the directive has
  been configured to support silenttokenrenewal (by specifying the
  silentTokenRedirectUrl parameter on the directive)  then it will call the
  'renewTokenSilently' method on the endpoint.  This code then sets a timeout
  to fire a minute before the token is expected to expire.  When the timeout
  is triggered some code executes that constructs a hidden iframe that calls
  out to the Identity Server's authorization endpoint with a prompt=none
  argument set, and an associated 5 second timeout to cleanup if the IDP does
  not respond within that timeframe. It also registers a listener for
  postMessage calls.  If the listener is triggered with some data of the
  string 'oauth2.silentRenewFailure' then it will broadcast the event
  'oauth2:authExpired' otherwise it will attempt to use the passed data
  as the location hash that would've been returned from the IDP, so the normal
  event broadcasing occurs (this should re-emit an 'oauth2:authSuccess' and restart
  the scheduled silent renewal again.)

Child iframe lifetime:
  Firstly this directive adds a new route 'silent-renew' to the route provider
  which has no template or controller, but as it is within the same template
  as the original oauth directive was nested we get to re-leverage the existing
  oauth configuration.  This means when setting up your OIDC client in your IP
  you will need to support at least two redirects, one for the outer-page (the
  existing routes) and a new one for the hidden inner frame
  (probably <scheme>://<host>/#/silent-renew).

  When requests come back in to the 'silent-renew' page they are treated as normal
  'post authentication' requests and the location is parsed as usual, if an
  'oauth2:authSuccess' event is emitted and capture by the child iframe then
  it will use 'postMessage' to send the silently-renewed token hash back to the
  parent frame, otherwise if an 'oauth2:authError' is captured by the child iframe
  then it will use 'postMessage' to send the string 'oauth2.silentRenewFailure'
  to the parent window.

ciaranj commented:
Hiyah, i've rebased this work against master again ;)

JamesRandall commented:
Thank you - I should be looking at this in the next week or so as I'm putting in place an STS that requires token renewal support and this library updating to do that.

JamesRandall commented:
Many many thanks for this. I've spent a couple of hours reviewing and testing it in a fairly complex Angular app and it works great.

In the morning I'm going to add a readme.md in the morning documenting the existing functionality and this addition and then I'll bump the bower version to 1.1.0 and release.

Thanks again - really appreciated.

ciaranj commented:
No worries, I've tested in my side too (admittedly more pre the recent rebase) and yes I've tried to document the behaviour completely in the commit message, so I definitely think documenting the functionality clearly would help, because it's a fiddly blighter ;)

JamesRandall commented:
And done. Documentation always takes longer than you think! Thanks again for your help.
